docs/geps/gep-04.rst

```diff
@@ -0,0 +1,210 @@
+
+.. code-block:: python
+
+    def soli_st_tu(
+        st_kind_freib_tu: FloatSeries,
+        anz_erwachsene_tu: IntSeries,
+        abgelt_st_tu: FloatSeries,
+        soli_st_params: dict,
+    ) -> FloatSeries:
+        ...
+
+
+would need to have ``kindergeld`` as a name for an input argument to request this
+:class:`pandas.Series`.
+
+Note that the type annotations (e.g. `FloatSeries`) indicate the expected type of each
+input and the output of a function.
+
+
+Directed Acyclic Graph
+~~~~~~~~~~~~~~~~~~~~~~
+
+The relationship between functions and their input variables is a graph where nodes are
+variables. These variables must either be present in the data supplied to GETTSIM or
+they are computed by functions. Edges are pointing from input variables to variables,
+which require them to be computed. See this `tutorial <../visualize_the_system.ipynb>`_
+for how to visualize this.
+
+The resulting structure is a special kind of graph, called a directed acyclic graph
+(DAG). It is directed because there are clearly inputs and outputs, i.e., a sense of
+direction. Acyclic means that there exist no path along the direction of the edges,
+where you start at some node and end up at the same node. Equivalently, a DAG has a
+topological ordering which is a sequence of nodes ordered from earlier to later in the
+sequence. The topological ordering is what defines the sequence in which the functions
+in the taxes and transfers system are evaluated. This ensures that the inputs are
+already computed before a function that requires them is called.
+
+When ``compute_taxes_and_transfers`` is called, GETTSIM builds a DAG based on three
+inputs provided by the user:
+
+ - A set of functions of the tax and transfer system, which consist of the ones
+   pre-implemented in GETTSIM and potentially user-written additional functions.
+ - The targets of interest.
+ - The input variables.
+
+ The DAG is then used to call all required functions in the right order and to
+ calculate the requested targets.
+
+
+Level of the DAG and limitations
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The DAG sketched above allows for great flexibility at any given point in time. It also
+allows the user to override any function easily. However, it does not allow the
+interface of functions to change over time, new functions to appear, or old ones to
+disappear.
+
+Some examples include:
+
+1. `arbeitsl_hilfe` being replaced by `arbeitsl_geld_2`
+1. `kinderbonus` being active only in a few years
+1. The introduction of `kinderzuschlag`
+1. Capital income entering `sum_brutto_eink` or not.
+
+The current solution handles the case of changing interfaces like this (example 4.):
+
+.. code-block:: python
+
+    if year < 2009:
+        functions["sum_brutto_eink"] = sum_brutto_eink_mit_kapital
+    else:
+        functions["sum_brutto_eink"] = sum_brutto_eink_ohne_kapital
+
+
+However, all functions will be present in all years in principle.
```

Comment:
This is not true. In the DAG, only the function `sum_brutto_eink` would be present in all years and only this argument can be used in other functions. We should probably describe this feature in this GEP in more detail.

The problem case is rather when we don't want to subsume the two concepts in a common name (as it is probably the case for `arbeitsl_hile`) or if a function does no longer exist (see `kinderzuschlag`).

Reply:
Sorry, my language was not precise enough.

You are right when it comes to the graph seen by the user. She would only see `sum_brutto_eink` indeed because this explicitly done in `set_up_policy_environment` / `load_reforms_for_date`. However, during the construction of the DAG as part of `compute_taxes_and_transfers`, the two underlying functions are present and one will just not be used. 

So my overall point is that it is pretty arbitrary right now what constitutes a policy that changes over time and thus ends up in `load_reforms_for_date` / `policy_functions` and what does not. I think we can and should be smarter than what we currently have.

Reply:
> So my overall point is that it is pretty arbitrary right now what constitutes a policy that changes over time and thus ends up in `load_reforms_for_date` / `policy_functions` and what does not. I think we can and should be smarter than what we currently have.

Thinking about it, I'd put it this way: 

1. As a principle, any parameter should be exactly specified as in the law. Do not use parameters derived from actual ones, as it is currently done for social insurance contributions or as it has been done before with Kinderzuschlag.
2. As a consequence, any formal representation of a policy will be close to the text of the law. There are limits to this as e.g. the income tax law is written in a horrible manner which you don't want to see formalized.

Now to the decision on whether we need a new function
3. If a parameter change can be implemented by using the current function, don't create a new function. 
4. If one or several **new parameters** are introduced and can be added to the existing framework by setting to some 'neutral value' (e.g. 0, 1, null) for years prior to its introduction, one should do that instead of creating a new function. An example for this is the introduction of the 'Zusatzfreibetrag' for health insurance contributions (#323). Applying this rule would call for single function for `alleinerziehend_freib_tu`, as it would be sufficient to set `eink_st_abzuege_params["alleinerziehenden_freibetrag_zusatz"]` to 0 until 2014 https://github.com/iza-institute-of-labor-economics/gettsim/blob/426dc9a6a755311876e5920f7b3289bc52fd0157/gettsim/taxes/zu_verst_eink/freibetraege.py#L40-L92 
5. If this is not possible, set up a new function.

Reply:
That's very much what I had in mind! Sounds good

Reply:
Sounds great! A couple of reactions:

> 1. Do not use parameters derived from actual ones, as it is currently done for social insurance contributions

At which point? In the parameters files? Yes. In the policy_params dict? Probably no. Both of those would be in the scope of GEP 3. Anything I am missing?

> 2. As a consequence, any formal representation of a policy will be close to the text of the law. There are limits to this as e.g. the income tax law is written in a horrible manner which you don't want to see formalized.

As a general rule, I agree. But I would not want to really constrain our implementations here. See, e.g., the proposal to change the calculation of social insurance contributions #332 

> 4. If one or several new parameters are introduced and can be added to the existing framework by setting to some 'neutral value' (e.g. 0, 1, null) for years prior to its introduction, one should do that instead of creating a new function. 

This is exactly what I'd hope we will be able to (largely) avoid by some clever re-design of how we deal with functions that change over time. Maybe that won't be possible, but it should be worth some brainstorming sessions.

Reply:
> At which point? In the parameters files? Yes. In the policy_params dict? Probably no. Both of those would be in the scope of GEP 3. Anything I am missing?

I was primarily referring to the parameter files.

> As a general rule, I agree. But I would not want to really constrain our implementations here. See, e.g., the proposal to change the calculation of social insurance contributions #332

Agreed, this is all guidelines, not the law😀. I'd just like to stress that aligning python code to legal code where possible has value in itself.

> This is exactly what I'd hope we will be able to (largely) avoid by some clever re-design of how we deal with functions that change over time. Maybe that won't be possible, but it should be worth some brainstorming sessions.

I'm just afraid this clever re-design migh lead to functions that have little in common with what can be read in the law. The last thing real-life

Reply:
> I'm just afraid this clever re-design might lead to functions that have little in common with what can be read in the law.

My hope would be just the opposite -- being able to write functions that look like the law at a particular point in time, which are not littered with parameters becoming effective only years later etc.. Let's see, you are definitely raising great points!

Reply:
So what's the plan? Keep the PR (and this comment) open until the brainstorming session (maybe during the workshop in April)?

Reply:
Yes. No need to put the current situation into writing, GEPs are exactly the process for channeling the brainstorming.

In the beginning, they just end up being larger and taking longer than what they will eventually be (GEP 5 would be a good example of the latter, though our process was not ideal).